### PR TITLE
Kubernetes integration

### PR DESCRIPTION
**Pull request by stephchen**
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
What does the PR do?

Testing
-------
How was this PR tested?

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/launch/runner/kubernetes.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            time.sleep(1)
+            timeout -= 1
+            job_response = self.batch_api.read_namespaced_job_status(name=self.name, namespace=self.namespace)
+        
+        if timeout == 0 or job_response.status.conditions[0].type != "Suspended":
+            raise LaunchError("Failed to suspend job {}. Check Kubernetes dashboard for more info.".format(self.name))
+
+    def cancel(self) -> None:
+        self.suspend()
+        # todo: is that right? suspension will already delete all active pods but suspended jobs can be resumed/uncancelled - should we delete the job
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:82]: I think if a user cancels a job it should also delete it.

File: wandb/sdk/launch/runner/kubernetes.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+class KubernetesRunner(AbstractRunner):
+    def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
+        kubernetes = get_module("kubernetes", "KubernetesRunner requires kubernetes to be installed")
+
+        resource_args = launch_project.resource_args.get("kubernetes")
+        if resource_args is None:
+            raise LaunchError("No Kubernetes resource args specified. Specify args via --resource-args with a JSON file or string under top-level key kubernetes")
+        
+        config_file = resource_args.get("config_file")   # kubeconfig, if None then loads default in ~/.kube
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:93]: Not sure this is needed if we're running in the kubernetes cluster already.

File: wandb/sdk/launch/runner/gcp_vertex.py [hunk 3]
```diff
@@ -103,12 +103,12 @@ def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
         gcp_staging_bucket = resource_args.get("gcp_staging_bucket")
         if not gcp_staging_bucket:
             raise LaunchError(
-                "Vertex requires a staging bucket for training and dependency packages in the same region as compute. You can specify a bucket with --resource-arg gcp_staging_bucket=<bucket>."
+                "Vertex requires a staging bucket for training and dependency packages in the same region as compute. Specify a bucket under key gcp_staging_bucket."
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/gcp_vertex.py:106]: We should put up a PR for CLI for these changes as well. (Not today)

File: wandb/sdk/launch/runner/kubernetes.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        else:
+            registry = resource_args.get("registry")
+            if registry is None:
+                # allow local registry usage for eg local clusters but throw a warning
+                wandb.termlog("Warning: No Docker registry specified. Image will be hosted on local registry, which may not be accessible to your training cluster.")
+
+            image_uri = construct_local_image_uri(launch_project)
+            if registry:
+                image_uri = os.path.join(registry, image_uri)
+            generate_docker_image(self._api, launch_project, image_uri, entry_point, docker_args, runner_type="kubernetes")
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:189]: I don't see any changes to generate_docker_image in this PR, does it do something different for kubernetes?

File: wandb/sdk/launch/runner/kubernetes.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            container_resources['requests'] = resource_args.get("resource_requests")
+        if resource_args.get('resource_limits'):
+            container_resources['limits'] = resource_args.get('resource_limits')
+        if container_resources:
+            containers[0]['resources'] = container_resources
+        # todo: args and env vars would be added here, need to figure out what kind of overrides we want
+
+        pod_spec['restartPolicy'] = resource_args.get('pod_restart_policy', 'Never')
+        if resource_args.get('pod_preemption_policy'):
+            pod_spec['preemptionPolicy'] = resource_args.get("pod_preemption_policy")
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:156]: are premptionPolicuy, nodeName and nodeSelectors optional?

File: wandb/sdk/launch/runner/kubernetes.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        # allow top-level namespace override, otherwise take namespace specified at the job level, or default
+        namespace = resource_args.get("namespace", job_metadata.get("namespace", "default"))
+
+        # name precedence: resource args override > name in spec file > generated name
+        job_metadata['name'] = resource_args.get("job_name", job_metadata.get('name'))
+        if not job_metadata.get('name'):
+            job_metadata['generateName'] = "launch-"
+
+        if resource_args.get('cluster_name'):
+            job_metadata['clusterName'] = resource_args.get("cluster_name")
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:127]: is the cluster name optional?

File: wandb/sdk/launch/runner/kubernetes.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        if resource_args.get('job_labels'):
+            job_metadata['labels'] = resource_args.get('job_labels')
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:129]: are job labels optional

File: wandb/sdk/launch/runner/kubernetes.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+            raise LaunchError("Resource overrides not supported for multiple containers. Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+
+        containers[0]['name'] = resource_args.get("container_name", containers[0].get('name', 'launch'))
+        container_resources = containers[0].get('resources', {})
+        if resource_args.get('resource_requests'):
+            container_resources['requests'] = resource_args.get("resource_requests")
+        if resource_args.get('resource_limits'):
+            container_resources['limits'] = resource_args.get('resource_limits')
+        if container_resources:
+            containers[0]['resources'] = container_resources
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:151]: same as other optional questions, if these aren't optional we should fail out when these aren't given

File: wandb/sdk/launch/runner/kubernetes.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+            time.sleep(5)
+        return status == "finished"     # todo: not sure if this (copied from aws runner) is the right approach? should we return false on failure
+
+    def get_status(self) -> Status:
+        job_response = self.batch_api.read_namespaced_job_status(name=self.name, namespace=self.namespace)
+        status = job_response.status
+        # todo: we only handle the 1 pod case. see https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs for multipod handling
+        if status.succeeded == 1:
+            return Status("finished")
+        elif status.failed == 1:
```
Comment (jlzhao27) [inline wandb/sdk/launch/runner/kubernetes.py:59]: nit: should use >=1 for failed, ive noticed k8s will retry your pod sometimes and end up with more than 1 failed containers.

File: wandb/sdk/launch/runner/kubernetes.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+)
+from ..utils import (
+    PROJECT_DOCKER_ARGS,
+    PROJECT_SYNCHRONOUS,
+)
+
+TIMEOUT = 5
+
+class KubernetesSubmittedRun(AbstractRun):
+    from kubernetes.client.api.batch_v1_api import BatchV1Api
```
Comment (jlzhao27) [inline wandb/sdk/launch/runner/kubernetes.py:32]: do we need to add the k8s package to some requirements.txt?

File: wandb/sdk/launch/runner/kubernetes.py (hunk 2 above)
Comment (vanpelt) [reply wandb/sdk/launch/runner/kubernetes.py:93]: We should likely be using RBAC and k8s service accounts.  Here's an example of loading the config from within the cluster: https://github.com/kubernetes-client/python/blob/master/examples/in_cluster_config.py

File: wandb/sdk/launch/runner/kubernetes.py (hunk 8 above)
Comment (vanpelt) [reply wandb/sdk/launch/runner/kubernetes.py:151]: These should be optional.  We might want to set the requests to the limits if only limits are specified.  That gives the pod a [guaranteed QoS](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed).

File: wandb/sdk/launch/runner/kubernetes.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+        image = resource_args.get("image")
+        if image:
+            if launch_project.docker_image and image != launch_project.docker_image:
+                raise LaunchError("Conflicting Docker images specified in launch and resource args.")
+            if len(containers) > 1:
+                raise LaunchError("Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+        user_provided_image = image or launch_project.docker_image
+        if user_provided_image:
+            pull_docker_image(user_provided_image)
```
Comment (vanpelt) [inline wandb/sdk/launch/runner/kubernetes.py:178]: Why are we pulling an image here?  k8s will pull the image when it needs to run it...

File: wandb/sdk/launch/runner/kubernetes.py (hunk 5 above)
Comment (vanpelt) [reply wandb/sdk/launch/runner/kubernetes.py:156]: Yep

File: wandb/sdk/launch/runner/kubernetes.py (hunk 7 above)
Comment (vanpelt) [reply wandb/sdk/launch/runner/kubernetes.py:129]: Yep, labels are optional.

Comment (vanpelt) [reply wandb/sdk/launch/runner/kubernetes.py:129]: But we might want to always add a label that can tell us it's one of "our jobs"

File: wandb/sdk/launch/runner/kubernetes.py (hunk 6 above)
Comment (vanpelt) [reply wandb/sdk/launch/runner/kubernetes.py:127]: I don't think we should ever be setting this.  We should likely always just use the current context or whatever cluster we're in with `config.load_incluster_config()`.  The only case the SDK would need this is if we wanted to support a k8s agent that can launch into multiple clusters which seems complicated and something we should definitely punt on.

File: wandb/sdk/launch/runner/kubernetes.py (hunk 2 above)
Comment (stephchen) [reply wandb/sdk/launch/runner/kubernetes.py:93]: doing this now

File: wandb/sdk/launch/runner/gcp_vertex.py (hunk 3 above)
Comment (stephchen) [reply wandb/sdk/launch/runner/gcp_vertex.py:106]: wait what do you mean

File: wandb/sdk/launch/runner/kubernetes.py (hunk 4 above)
Comment (stephchen) [reply wandb/sdk/launch/runner/kubernetes.py:189]: nope although after this PR i need to add generalized command/entrypoint functionality

File: wandb/sdk/launch/runner/kubernetes.py (hunk 8 above)
Comment (stephchen) [reply wandb/sdk/launch/runner/kubernetes.py:151]: these are optional

File: wandb/sdk/launch/runner/kubernetes.py (hunk 10 above)
Comment (stephchen) [reply wandb/sdk/launch/runner/kubernetes.py:32]: in general we don't want to require that all wandb users install k8s — this call should be wrapped in `use_module`, good catch

File: wandb/sdk/launch/runner/kubernetes.py (hunk 11 above)
Comment (stephchen) [reply wandb/sdk/launch/runner/kubernetes.py:178]: yup removing these (also from aws/gcp)

File: wandb/sdk/launch/runner/gcp_vertex.py (hunk 3 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/runner/gcp_vertex.py:105]: These are good fixes to have in master. There's a few PRs in this push that we'll end up wanting in master as well. These changes included

File: wandb/sdk/launch/runner/kubernetes.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+class KubernetesRunner(AbstractRunner):
+    def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:  # noqa: C901
+        kubernetes = get_module(  # noqa: F811
+            "kubernetes", "KubernetesRunner requires kubernetes to be installed"
+        )
+
+        resource_args = launch_project.resource_args.get("kubernetes")
+        if resource_args is None:
+            raise LaunchError(
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes.py:105]: Since this will work without any specified args should we remove this?

**Review by KyleGoyette — APPROVED** [review]
Pending the one comment I left about removing the error if someone provides no resrouce args, this lgtm

**Pull request merged**